traffic_gym.py

```diff
@@ -1024,6 +1040,61 @@ def _draw_lanes(self, surface, mode='human', offset=0):
                 draw_line(surface, colours['r'], (0, lane['min'] + m), (sw, lane['min'] + m), 1)
                 draw_line(surface, colours['r'], (0, lane['max'] + m), (sw, lane['max'] + m), 1)
 
+    def draw_trajectory(self, v):
+        pad = 0
+        if self.trajectory_image is None:
+            self.trajectory_image = np.zeros((self.screen_size[0], self.screen_size[1], 4))
+        trajectory = v._trajectory
+        # remove nan
+        real_len = 0
+        for i in range(len(trajectory) - 1, 0, -1):
+            if not math.isnan(trajectory[i, 0]) and not math.isnan(trajectory[i, 1]):
+                real_len = i + 1
+                break
+        # remove waittime
+        points = []
+        points.append(trajectory[0, :])
+        for i in range(real_len):
+            if math.sqrt(pow(trajectory[i, 0] - points[-1][0], 2) + pow(trajectory[i, 1] - points[-1][1], 2)) > 5:
+                points.append(trajectory[i, :])
+        points = np.array(points)
+        curve = bezier.Curve(points.T, degree=len(points) - 1)
+        s_vals = np.linspace(0.0, 1.0, 5 * (len(points) - 1))
+        points = curve.evaluate_multi(s_vals).T
+
+        for i in range(len(points) - 1):
+            x, y = points[i, 0], points[i, 1]
+            p = (int(x), int(y))
+            d = points[i + 1, :] - points[i, :]
+            # HSV 2 RGB
+            rad = math.atan(d[1] / (d[0] + 1e-6))
+            rad = (rad + math.pi / 2) / math.pi
+            if d[0] > 0:
+                H = rad * 0.5
+            else:
+                H = rad * 0.5 + 0.5
+            S = 1
+            V = 1
+            RGBcolor = hsv_to_rgb((H, S, V))
+            self.trajectory_image[p[0] - pad:p[0] + pad + 1, p[1] - pad:p[1] + pad + 1, 0] += RGBcolor[0]
+            self.trajectory_image[p[0] - pad:p[0] + pad + 1, p[1] - pad:p[1] + pad + 1, 1] += RGBcolor[1]
+            self.trajectory_image[p[0] - pad:p[0] + pad + 1, p[1] - pad:p[1] + pad + 1, 2] += RGBcolor[2]
+            self.trajectory_image[p[0] - pad:p[0] + pad + 1, p[1] - pad:p[1] + pad + 1, 3] += 1
+
+    def get_final_trajectory(self):
+        threshold = 0.95
+        self.trajectory_image = self.trajectory_image[:, :, 0:3] / np.expand_dims(self.trajectory_image[:, :, 3] + 1e-6,
+                                                                                  axis=2)
+        hsv = rgb_to_hsv(self.trajectory_image)
+        s = hsv[:, :, 1]
+        s[s < threshold] = 0
+        hsv[:, :, 1] = s
+        v = hsv[:, :, 2]
+        v[v > 0] = 1
+        hsv[:, :, 2] = v
+        self.trajectory_image = hsv_to_rgb(hsv)
+        plt.imsave("1trajectory.jpg", np.transpose(self.trajectory_image, (1, 0, 2)))
```

Comment:
Is the `1` a typo?

Reply:
It is a symbol for original lane map. I will remove it.